File: src/main/java/io/github/demianli/projectmcp/gh/Remedy.java
package io.github.demianli.projectmcp.gh;

/**
 * What a caller should <em>do next</em> about a failure. Failures are classified by the
 * action available, not by the cause. See docs/design.md#failure-contract.
 */
public enum Remedy {

    /**
     * Try the same call again. May carry a wait, in which case do not retry before it — the
     * wait comes only from a rate limit, either one {@code gh} has named or this Server's own
     * limit on writes, never from a timeout.
     */
    RETRY,

    /**
     * A write whose result this Server could not read. Look whether it landed by
     * {@code list_issue_comments}, then decide: if the comment is there, success; if not,
     * {@link #RETRY}. See docs/design.md#writes.
     */
    CHECK_BEFORE_RETRY,

    /** The call cannot succeed as written. Change the arguments. */
    FIX_REQUEST,

    /** Nothing the caller can change. A human has to fix the environment. */
    ASK_OPERATOR,

    /**
     * {@code gh} failed in a way this Server does not recognise. What travels alongside is
     * all there is to go on — usually {@code gh}'s stderr verbatim, and on the one path
     * where {@code gh} produced none, a read whose output could not be read back, a Java
     * exception string standing in that slot instead.
     */
    UNKNOWN
}

File: src/tools/remedy.ts
/**
 * 呼叫者下一步該做什麼。
 *
 * 依「可以採取的行動」分類，不依成因——兩個成因不同、但下一步相同的失敗，共用同一個
 * Remedy。工具失敗走 `isError: true` 的結果而不是 JSON-RPC error，模型才看得到它，
 * 也才有機會改參數再試。
 */
export const Remedy = {
  /** 原樣再呼叫一次。 */
  Retry: 'RETRY',
  /** 寫入的結果無法讀取：先確認是否已經寫入，再決定要不要重送。 */
  CheckBeforeRetry: 'CHECK_BEFORE_RETRY',
  /** 照原樣呼叫不可能成功，必須改參數。 */
  FixRequest: 'FIX_REQUEST',
  /** 呼叫者改不了任何事，要由人修正環境。 */
  AskOperator: 'ASK_OPERATOR',
  /** 無法辨識的失敗。 */
  Unknown: 'UNKNOWN',
} as const;

export type Remedy = (typeof Remedy)[keyof typeof Remedy];
